stdlib/public/core/StringSwitch.swift

```diff
@@ -0,0 +1,96 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+// This file contains compiler intrinsics for optimized string switch
+// implementations. All functions and types declared in this file are not
+// intended to be used by switch source directly.
+
+/// The compiler intrinsic which is called to lookup a string in a table
+/// of static string case values.
+@_semantics("stdlib_binary_only")
+@_semantics("findStringSwitchCase")
+public // COMPILER_INTRINSIC
+func _findStringSwitchCase(
+  cases: [StaticString],
+  string: String) -> Int {
+
+  for (idx, s) in cases.enumerated() {
+    if String(_builtinStringLiteral: s.utf8Start._rawValue,
+              utf8CodeUnitCount: s._utf8CodeUnitCount,
+              isASCII: s.isASCII._value) == string {
+      return idx
+    }
+  }
+  return -1
+}
+
+public // used by COMPILER_INTRINSIC
+struct _OpaqueStringSwitchCache {
+  var a: Builtin.Word
+  var b: Builtin.Word
+}
+
+internal typealias _StringSwitchCache = Dictionary<String, Int>
+
+internal struct _StringSwitchContext {
+  let cases: [StaticString]
+  let cachePtr: UnsafeMutablePointer<_StringSwitchCache>
+}
+
+/// The compiler intrinsic which is called to lookup a string in a table
+/// of static string case values.
+///
+/// The first time this function is called, a cache is built and stored
+/// in \p cache. Consecutive calls use the cache for faster lookup.
```

Comment:
Does this mean that subsequent calls can ignore the cases parameter, or might the cases content change between calls?

Reply:
If they won't change then please mention the guarantee in the comments. Then, LGTM

Reply:
Yeah, the cases content must not change. Although the current implementation does not use the cases parameter in subsequent calls, a future implementation might do, e.g. the built cache may reference strings in the cases array.